flycheck.el

```diff
@@ -2785,28 +2870,47 @@ buffer."
   ;; Save and restore the match data, as recommended in (elisp)Change Hooks
   (save-match-data
     (when flycheck-mode
-      ;; The buffer was changed, thus clear the idle timer
-      (flycheck-clear-idle-change-timer)
       (if (string-match-p (rx "\n") (buffer-substring beg end))
           (flycheck-buffer-automatically 'new-line 'force-deferred)
-        (setq flycheck-idle-change-timer
+        (flycheck--clear-idle-trigger-timer)
+        (cl-pushnew 'idle-change flycheck--idle-trigger-conditions)
+        (setq flycheck--idle-trigger-timer
               (run-at-time flycheck-idle-change-delay nil
-                           #'flycheck--handle-idle-change-in-buffer
+                           #'flycheck--handle-idle-trigger
                            (current-buffer)))))))
 
-(defun flycheck--handle-idle-change-in-buffer (buffer)
-  "Handle an expired idle timer in BUFFER since the last change.
-This thin wrapper around `flycheck-handle-idle-change' is needed
-because some users override that function, as described in URL
-`https://github.com/flycheck/flycheck/pull/1305'."
-  (when (buffer-live-p buffer)
-    (with-current-buffer buffer
-      (flycheck-handle-idle-change))))
-
-(defun flycheck-handle-idle-change ()
-  "Handle an expired idle timer since the last change."
-  (flycheck-clear-idle-change-timer)
-  (flycheck-buffer-automatically 'idle-change))
+(defvar flycheck--last-buffer (current-buffer)
+  "The current buffer or the buffer that was previously current.
+This is usually equal to the current buffer, unless the user just
+switched buffers.  After a buffer switch, it is the previous
+buffer.")
+
+(defun flycheck-handle-buffer-switch ()
+  "Handle a possible switch to another buffer.
+
+If a buffer switch actually happened, schedule a syntax check."
+  ;; Switching buffers here is weird, but unfortunately necessary.  It
+  ;; turns out that `with-temp-buffer' triggers
+  ;; `buffer-list-update-hook' twice, and the value of
+  ;; `current-buffer' is bogus in one of those triggers (the one just
+  ;; after the temp buffer is killed).  If we rely on the bogus value,
+  ;; Flycheck will think that the user is switching back and forth
+  ;; between different buffers during the `with-temp-buffer' call
+  ;; (note: two different normal buffers, not the current buffer and
+  ;; the temp buffer!), and that would trigger spurious syntax checks.
+  ;; It seems that reading (car (buffer-list)) gets us the correct
+  ;; current buffer in all important real-life situations (although it
+  ;; doesn't necessarily catch uses of `set-buffer').
+  (with-current-buffer (car (buffer-list))
+    (unless (equal flycheck--last-buffer (current-buffer))
+      (setq flycheck--last-buffer (current-buffer))
+      (when flycheck-mode
+        (flycheck--clear-idle-trigger-timer)
+        (cl-pushnew 'idle-buffer-switch flycheck--idle-trigger-conditions)
```

Comment:
Can you elaborate on why we accumulate the conditions here?  It seems this is necessary because we have only one function `flycheck--handle-idle-trigger` now.  Is that it?

Reply:
Basically, it is so that I can handle the case that I mention in my comment below—an idle change followed by a buffer switch, when `flycheck-buffer-switch-check-intermediate-buffers` is nil. We want to be able to say "cancel any syntax check that was due to a buffer switch, but don't cancel any syntax check that was due to an edit". If I use a buffer-local list of conditions, it is easy to handle that:

1. When there is an edit, add `idle-change` to the list
2. When there is a buffer switch, add `idle-buffer-switch` to the list
3. When it comes time to perform the syntax check, remove `idle-buffer-switch` from the list unless the user is in the same buffer as they were in when the syntax check was scheduled

I think that handling this correctly will always require the use of an additional variable. In my previous attempt, that variable was `flycheck--force-checking-intermediate-buffers`.

Reply:
> "cancel any syntax check that was due to a buffer switch, but don't cancel any syntax check that was due to an edit"

Ok, that's indeed a case I had not considered.  Thanks for pointing it out.

> I think that handling this correctly will always require the use of an additional variable.

I concur.  Although, for this particular case, is it necessary that we keep track of all the "causes" for triggers?  I.e., since there is only one active timer, just checking if the last timer was triggered by a buffer-switch is enough to be able to cancel it if `flycheck-buffer-swithc-check-intermediate-buffers` is nil.

> When it comes time to perform the syntax check, remove idle-buffer-switch from the list unless the user is in the same buffer as they were in when the syntax check was scheduled

Shouldn't we always clear the list of causes when we actually run the check instead?  And what purpose does this condition serve?  It doesn't seem necessary to handle the case spelt out above.

Reply:
>  just checking if the last timer was triggered by a buffer-switch is enough to be able to cancel it if `flycheck-buffer-switch-check-intermediate-buffers` is nil.

I don't think that's the case, and I have a more complex example to demonstrate why. Suppose the user makes an edit in the buffer. Then the last timer was triggered by an edit. But now suppose that, before the timer triggers, the user quickly switches from that buffer to a different one, then back. We have to reset the timer, and now the last timer was triggered by a buffer-switch. Finally, suppose that before *that* timer expires, the user again switches to another buffer. We notice that the last timer was triggered by a buffer-switch, and cancel it. But now no timer will ever trigger in that buffer: we have forgotten about the edit.

I think you *have* to keep track of each condition separately. A toggle, or a single condition name, is simply not enough.

> Shouldn't we always clear the list of causes when we actually run the check instead? 

Indeed, we actually do that:

<https://github.com/flycheck/flycheck/pull/1308/files#diff-374cdd9563b826c712b1dc6a373a0bc1R2825>

You're probably wondering why we *also* do it in `flycheck--handle-idle-trigger`. That's because we want to clear out the list of conditions even if `flycheck-check-syntax-automatically` decides not to perform a syntax check. Otherwise the list will just accumulate over time.

> It doesn't seem necessary to handle the case spelt out above.

Does my more complex example in the first paragraph answer this question?

Reply:
> But now no timer will ever trigger in that buffer: we have forgotten about the edit.

Right!  And if we keep track of both causes, we would not cancel the last timer, which would eventually run a check for `idle-change`.  That justifies the variable.

Although it's arguable whether the added complexity is worth it.  If your check is not triggered after this frantic buffer switching, you would just switch back to the original buffer to trigger an `idle-buffer-switch` check.

I'll take another look at the PR with these clarifications in mind.  Thank you again for spelling things out.

However, now I'm wondering what happens if you do *not* have `idle-buffer-switch` in `flycheck-check-syntax-automatically`.  We do not check for the value of that variable, so the exact same events will unfold, right?  We will reset the timer and delay the `idle-change` by `flycheck-idle-buffer-switch-delay`.  Although we will eventually run the check, it will have been delayed unexpectedly.  This seems to contradict the docstring:
```
This variable has no effect if `idle-buffer-switch' is not
contained in `flycheck-check-syntax-automatically'
```
We can probably add a guard that looks for `idle-buffer-switch` to behave as that docstring states.  I'd rather not go back to the two timers implementation.

Reply:
> We can probably add a guard

Agreed; added.